### DISCUSSION TITLE
Changing the Cahn-Hilliard kernel

### DISCUSSION THREAD

**Discussion by souzanha** [idaholab/moose#17302] [Q&A Modules: Phase field] [answered]
Hi moose-team, 
I want to couple a variable (concentration) into the Cahn-Hilliard kernel, c(1-c)*M*grad(mu). I'm not entirely sure I'm doing it correctly though. For example, the section where you iterate over all coupled variables: 

  _// Iterate over all coupled variables
  for (unsigned int i = 0; i < _n_args; ++i)
    _dmobdarg[i] = &getMaterialPropertyDerivative<T>(_mob_name, i);
}_

Is it enough to do it this way and only include _c[_qp] in the residual and jacobian expressions? What am I missing? : 

input file: 
  [./wres]
    type=SplitCHWResT
    variable = mu
    mob_name = 1.76e-23
    **coupled_conc = c**  
  [../]

SplitCHWResBaseT-file:
private:
 const VariableValue & _c;

 _c(coupledValue("coupled_conc"))

RESIDUAL
template <typename T>
Real
SplitCHWResBaseT<T>::computeQpResidual()
{
  return _c[_qp] * (1-_c[_qp]) * _mob[_qp] * _grad_w[_qp] * _grad_test[_i][_qp];
}

JACOBIAN
template <typename T>
Real
SplitCHWResBaseT<T>::computeQpWJacobian()
{
  return _c[_qp] * (1-_c[_qp]) * _mob[_qp] * _grad_phi[_j][_qp] * _grad_test[_i][_qp];
}

  params.addCoupledVar(
      "coupled_conc", "Coupled concentration (if not specified kernel variable will be used)");
  return params;

**Comment by yucoutinho** [comment, marked as answer]
hi @souzanha,

If I understood correctly, you are trying to make the mobility composition dependent as
M(c) = c(1-c)M<sub>const</sub>

Instead of modifying the Cahn-Hilliard kernel, you can create a material,

```
[Materials]
  [./mobility]
    type = DerivativeParsedMaterial
    f_name = mob
    function = 'c*(1-c)*M'
    args = c
    constant_names =  M
    constant_expressions = 1.76e-23
    derivative_order = 1
  [../]
[]
```

and then use the standard kernel

```
[Kernels]
  [./chw_res]
    variable = w
    type = SplitCHWRes
    mob_name = mob
  [../]
[]
```
Check this [example](https://github.com/idaholab/moose/blob/next/modules/phase_field/tutorials/spinodal_decomposition/s4_mobility.i).

Cheers

**Reply by souzanha** [reply]
Hi @yucoutinho! 
Thank you for the suggestion. That is definitely a much better approach.
